src/lib/replication/repl_dev/raft_state_machine.cpp

```diff
@@ -0,0 +1,510 @@
+#include <iomgr/iomgr_timer.hpp>
+#include <sisl/logging/logging.h>
+#include <sisl/fds/utils.hpp>
+#include <sisl/fds/vector_pool.hpp>
+
+#include "repl_dev/raft_state_machine.h"
+#include "repl_dev/raft_repl_dev.h"
+
+SISL_LOGGING_DECL(replication)
+
+namespace homestore {
+
+RaftStateMachine::RaftStateMachine(RaftReplDev& rd) : m_rd{rd}, m_group_id{m_rd.group_id()} {
+    m_success_ptr = nuraft::buffer::alloc(sizeof(int));
+    m_success_ptr->put(0);
+}
+
+raft_buf_ptr_t RaftStateMachine::pre_commit_ext(nuraft::state_machine::ext_op_params const& params) {
+    // Leader precommit is processed in next callback, since lsn would not have been known to repl layer till we get
+    // the next callback.
+    if (!m_rd.is_leader()) {
+        int64_t lsn = s_cast< int64_t >(params.log_idx);
+        raft_buf_ptr_t data = params.data;
+
+        RD_LOG(DEBUG, "pre_commit: {}, size: {}", lsn, data->size());
+        repl_req_ptr_t rreq = lsn_to_req(lsn);
+
+        m_rd.m_listener->on_pre_commit(rreq->lsn, rreq->header, rreq->key, rreq);
+    }
+    return m_success_ptr;
+}
+
+void RaftStateMachine::after_precommit_in_leader(nuraft::raft_server::req_ext_cb_params const& params) {
+    repl_req_ptr_t rreq = repl_req_ptr_t(r_cast< repl_req_ctx* >(params.context));
+    link_lsn_to_req(rreq, int64_cast(params.log_idx));
+    m_rd.m_listener->on_pre_commit(rreq->lsn, rreq->header, rreq->key, rreq);
+}
+
+raft_buf_ptr_t RaftStateMachine::commit_ext(nuraft::state_machine::ext_op_params const& params) {
+    int64_t lsn = s_cast< int64_t >(params.log_idx);
+    raft_buf_ptr_t data = params.data;
+
+    RD_LOG(DEBUG, "apply_commit: {}, size: {}", lsn, data->size());
+
+    repl_req_ptr_t rreq = lsn_to_req(lsn);
+    if (m_rd.is_leader()) {
+        // This is the time to ensure flushing of journal happens in leader
+        if (m_rd.m_data_journal->last_durable_index() < uint64_cast(lsn)) { m_rd.m_data_journal->flush(); }
+        rreq->state.fetch_or(uint32_cast(repl_req_state_t::LOG_FLUSHED));
+    }
+    if (rreq->state.load() & uint32_cast(repl_req_state_t::DATA_WRITTEN)) {
+        m_rd.report_committed(rreq);
+        m_lsn_req_map.erase(rreq->lsn);
+    }
+    return m_success_ptr;
+}
+
+uint64_t RaftStateMachine::last_commit_index() { return uint64_cast(m_rd.get_last_commit_lsn()); }
+
+void RaftStateMachine::propose_to_raft(repl_req_ptr_t rreq) {
+    uint32_t entry_size = sizeof(repl_journal_entry) + rreq->header.size() + rreq->key.size() +
+        (rreq->value.size ? rreq->local_blkid.serialized_size() : 0);
+    rreq->alloc_journal_entry(entry_size, true /* raft_buf */);
+    rreq->journal_entry->code = (rreq->value.size) ? journal_type_t::HS_LARGE_DATA : journal_type_t::HS_HEADER_ONLY;
+    rreq->journal_entry->user_header_size = rreq->header.size();
+    rreq->journal_entry->key_size = rreq->key.size();
+    rreq->journal_entry->dsn = rreq->dsn();
```

Comment:
forget to set rreq->journal_entry->value_size?

Reply:
Will check that and add if it is missing. Thanks for the catch.